File: 0229-majority-element-ii/0229-majority-element-ii.go
func majorityElement(nums []int) []int {
    c1, c2, f1, f2 := 0, 0, 0, 0
	for i := 0 ; i < len(nums) ; i++ {
		if f1 == 0 && nums[i] != c2{
			c1 = nums[i]
			f1 = 1
		} else if f2 == 0 && nums[i] != c1 {
			c2 = nums[i]
			f2 = 1
		} else if nums[i] == c1 {
			f1++
		} else if nums[i] == c2 {
			f2++
		} else {
			f1--
			f2--
		}
	}
	f1, f2 = 0, 0
	for i := 0 ; i < len(nums) ; i++ {
		if nums[i] == c1{
			f1++
		} else if nums[i] == c2 {
			f2++
		}
	}
	ret := []int{}
	n := len(nums)

	if f1 > n/3 {
		ret = append(ret, c1)
	}
	if f2 > n/3 {
		ret = append(ret, c2)
	}

	return ret
	
}

File: 0229-majority-element-ii/0229-majority-element-ii.py
# boyer-moore majority voting algorithm    
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        c1, c2, f1, f2 = 0, 0, 0, 0

        for n in nums:
            if f1 == 0 and n != c2:
                c1 = n
                f1 = 1
            elif f2 == 0 and n != c1:
                c2 = n
                f2 = 1
            elif n == c1:
                f1 += 1
            elif n == c2:
                f2 += 1
            else:
                f1 -= 1
                f2 -= 1

        f1, f2 = 0, 0

        for n in nums:
            if n == c1:
                f1 += 1
            elif n == c2:
                f2 += 1

        ret = []

        if f1 > len(nums)//3:
            ret.append(c1)
        if f2 > len(nums)//3:
            ret.append(c2)
        return ret 

            
                

            